Pendulum/Data/Export.swift

```diff
@@ -32,19 +64,148 @@ struct ExportedPenPal: Codable {
-    let archived: Bool
     let events: [ExportedEvent]
+    let archived: Bool
+    
+    init(from: PenPal, in context: NSManagedObjectContext) {
+        self.id = from.id ?? UUID()
+        self.name = from.wrappedName
+        self.initials = from.wrappedInitials
+        self.notes = from.notes
+        self.archived = from.archived
+        self.events = from.events(from: context).map { ExportedEvent(from: $0) }
+    }
+    
+    func loadImage(fromArchive archiveDirectory: URL) -> Data? {
+        do {
+            return try Data(contentsOf: archiveDirectory.appendingPathComponent("penpal-\(id.uuidString)").appendingPathExtension("png"))
+        } catch {
+            appLogger.debug("Could not load contact image \(id) from archive \(archiveDirectory)")
+        }
+        return nil
+    }
+    
 }
 
 struct ExportedStationery: Codable {
     let id: UUID
     let type: String
     let value: String
+    
+    init(from: Stationery) {
+        self.id = from.id ?? UUID()
+        self.type = from.wrappedType
+        self.value = from.wrappedValue
+    }
+    
+}
+
+struct ImportResult {
+    let stationeryCount: Int
+    let penPalCount: Int
+    let eventCount: Int
+    let photoCount: Int
 }
 
 struct Export: Codable {
     let penpals: [ExportedPenPal]
     let stationery: [ExportedStationery]
+    var context: NSManagedObjectContext? = nil
+    
+    enum CodingKeys: CodingKey {
+        case penpals, stationery
+    }
+    
+    init(from context: NSManagedObjectContext) {
+        self.penpals = PenPal.fetchAll(from: context).map { ExportedPenPal(from: $0, in: context) }
+        self.stationery = Stationery.fetch(from: context).map { ExportedStationery(from: $0) }
+        self.context = context
+    }
+    
+    func asJSON() throws -> Data {
+        let encoder = JSONEncoder()
+        return try encoder.encode(self)
+    }
+    
+    var name: String {
+        let dateFormatter = DateFormatter()
+        dateFormatter.locale = Locale(identifier: "en_US")
+        dateFormatter.dateFormat = "yyyy-MM-dd"
+        return "PendulumExport-\(dateFormatter.string(from: .now))"
+    }
+    
+    func export() throws -> URL {
+        
+        let fileName = name
+        
+        // Create temporary directory
+        let directoryURL = FileManager.default.temporaryDirectory.appendingPathComponent(fileName)
+        appLogger.debug("\(directoryURL)")
+        try? FileManager.default.removeItem(atPath: directoryURL.path(percentEncoded: false))
+        try FileManager.default.createDirectory(atPath: directoryURL.path(percentEncoded: false), withIntermediateDirectories: true)
+        
+        // Save JSON data
+        let jsonFilePath = directoryURL.appendingPathComponent("data").appendingPathExtension("json")
+        appLogger.debug("\(jsonFilePath)")
+        let jsonData = try self.asJSON()
+        try jsonData.write(to: jsonFilePath)
+        
+        // Save PenPal photos
+        if let context {
```

Comment:
If this is nil for some reason, it's probably a good idea to inform the user their photos weren't exported? Or maybe make it not optional so that can never happen?

Reply:
This is optional, because it's not part of the `Codable` stuff, and when I decode the data on re-import, I can't set it (see line 194).

It would only be nil if I ever created an `Export` any way other than with `.init(from:)`, which I don't… but I suppose I could forget at a later date, though I don't see that happening.

Unless there's a way to set it when decoding the JSON, perhaps with the JSON `userInfo` stuff, but I didn't really want to write my own `decode` implementation…

Reply:
Oh okay, I would suggest passing it as a non-optional parameter whenever needed instead of storing a local property in that case

Reply:
I did have that originally, but it felt weird to do:

```swift
let exportData = Export(from: context)
let exportURL = try? exportData.export(from: context)
```

Because `exportData` was generated from `context` so why "should" I have to pass it again

Reply:
That looks better IMO but I get it

Reply:
As it's only used for photos, perhaps I'll change it to `exportData.export(usingPhotosFrom: context)` to make it clear that everything else is already included 🤔